crates/bevy_ecs/src/component.rs

```diff
@@ -261,7 +279,69 @@ impl ComponentInfo {
 
     /// Create a new [`ComponentInfo`].
     pub(crate) fn new(id: ComponentId, descriptor: ComponentDescriptor) -> Self {
-        ComponentInfo { id, descriptor }
+        ComponentInfo {
+            id,
+            descriptor,
+            hooks: ComponentHooks::default(),
+        }
+    }
+
+    /// Register a [`ComponentHook`] that will be run when this component is added to an entity
+    ///
+    /// Will panic if the component already has an `on_add` hook
+    pub fn on_add(&mut self, hook: ComponentHook) -> &mut Self {
+        assert!(
```

Comment:
We could definitely allow this but it could cause scenarios where plugins stop working as expected as a user downstream replaces an important hook on a foreign type. (runtime warnings are easily ignored)

Observers don't face the same uniqueness constraint and can be spawned and despawned for use in different scenes so I would argue they suit this use case better than hooks, but I'm not entirely opposed.

Reply:
Probably best to just push either out and switch if needed based on feedback.